llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp

```diff
@@ -1009,8 +1015,16 @@ static void simplifyRecipes(VPlan &Plan, LLVMContext &Ctx) {
       Plan.getEntry());
   VPTypeAnalysis TypeInfo(Plan.getCanonicalIV()->getScalarType(), Ctx);
   for (VPBasicBlock *VPBB : VPBlockUtils::blocksOnly<VPBasicBlock>(RPOT)) {
-    for (VPRecipeBase &R : make_early_inc_range(*VPBB)) {
-      simplifyRecipe(R, TypeInfo);
+    // Populate a Worklist, as simplifyRecipe might return a new recipe that we
+    // need to re-process.
+    SmallVector<VPRecipeBase *> Worklist;
+    for (auto &R : VPBB->getRecipeList())
+      Worklist.push_back(&R);
+
+    while (!Worklist.empty()) {
+      VPRecipeBase *R = Worklist.pop_back_val();
+      for (VPRecipeBase *Cand : simplifyRecipe(*R, TypeInfo))
+        Worklist.push_back(Cand);
```

Comment:
Thanks for splitting this off. A next step could be for the work-list to also re-queue users of a simplified value for further simplification. With the current interface, we would have to create and return a number of temporary vectors for each simplification. Just wondering what your thoughts are about adding the the worklist in `simplifyRecipe`

Reply:
Hm, not sure I understand why we'd want to have the worklist in `simplifyRecipe`: I was thinking that we could keep `simplifyRecipe` simple, simply handling different patterns like the way it's done in InstCombine. I'm, in fact, quite fond of the approach I've taken.

Reply:
Is the idea that we pass a recipe to `simplifyRecipe`, which then maximally simplifies it, by running over the same code multiple times using a worklist every time a new recipe is introduced by a simplification? Wouldn't it be preferable to instead call `simplifyRecipe` multiple times, and maintain the worklist in `simplifyRecipes`, in the interest of having `simplifyRecipe` do one thing? I would imagine that if we took the former route, `simplifyRecipe` would be called something like `maximallySimplifyRecipe` instead?

Reply:
> Thanks for splitting this off. A next step could be for the work-list to also re-queue users of a simplified value for further simplification. With the current interface, we would have to create and return a number of temporary vectors for each simplification. Just wondering what your thoughts are about adding the the worklist in `simplifyRecipe`

Regarding re-queueing user: the only users affected are those of the root recipe, via RAUW. They should be visited, w/o re-queueing, as long as (a) recipes are simplified in def-use order - as in from first to last in block, and (b) each original recipe is simplified completely, possibly including simplifying new recipes recursively, before moving to the next. Right?

Reply:
They should be visited, w/o re-queueing, as long as (a) recipes are simplified in def-use order - as in from first to last in block,

Isn't the current order a reverse-post-order-traversal? When stepping through the code, I see that recipes are simplified from the last instruction in the block to the first instruction in the block.

Will make_early_inc_range automatically pick up RAUW'd definitions? I thought it would skip modified (or replaced) definitions altogether.

and (b) each original recipe is simplified completely, possibly including simplifying new recipes recursively, before moving to the next. Right?

I'm a little worried that we're painting ourselves into a corner by imposing the condition that `simplifyRecipe` only ever returns a single newly-introduced (sub)recipe.